R/utils.R

```diff
@@ -291,6 +291,20 @@ check_sampler_transitions_treedepth <- function(post_warmup_sampler_diagnostics,
   }
 }
 
+check_bfmi <- function(post_warmup_sampler_diagnostics) {
+  if (!is.null(post_warmup_sampler_diagnostics)) {
+    energy <- posterior::extract_variable_matrix(post_warmup_sampler_diagnostics, "energy__")
+    ebfmi <- apply(energy, 2, function(x) {
+      (sum(diff(x)^2)/length(x))/var(x)
+    })
+    if (any(ebfmi < .3)) {
+      message(sum(ebfmi < .3), " of ", length(ebfmi) , " chains had estimated Bayesian fraction
+      of missing information(E-BFMI) less than 0.3, which may indicate poor exploration of the 
```

Comment:
How about a `threshold` argument with a default of 0.3 instead of fixing it at 0.3?

Reply:
I'm not sure but maybe we should use the RStan threshold of 0.2. It sounds like the CmdStan threshold is maybe too conservative: 

> We set the E-BFMI warning threshold at 0.3 based on some empirical heuristics but that may have been a bit conservative given how robustly dynamic HMC handles even extremely heavy-tailed distributions. 

from https://discourse.mc-stan.org/t/fixing-bfmi-was-low-problems/658/2. And I'm not sure we have sufficiently good recommendations to justify throwing the warning more often. 

That said, when we add `fit$diagnose()` we can let the user pick the thresholds and be as conservative as they want. Here I'm just thinking about when the warning should automatically happen.

Reply:
This all makes sense. On the other hand, it could be confusing for the user to get different default behavior from however we implement this diagnostic on the R6 object and `$cmdstan_diagnose()`.  Thoughts @jgabry?

Reply:
Hmm, that's a good point. I would normally lean towards matching CmdStan in that case, however pretty soon nobody will need `cmdstan_diagnose()` anymore (and maybe we should consider deprecating when that's the case) because all of the same checks will be available directly in R (and the R implementations will be preferred since they'll be accessible even after `as_cmdstan_fit()` and even if the CSV files aren't present). So maybe in anticipation of that we should go with the threshold we prefer and not worry about matching cmdstan_diagnose? 

@jsocolar and @rok-cesnovar Do you have strong preferences here?

Reply:
I definitely don't have preferences. Additional food for thought regarding deprecation of `cmdstan_diagnose()`:
* It's currently the only way to get check the sampler diagnostics and the chain diagnostics (rhat, etc) all at once.
* Is `cmdstan_diagnose()` nontrivially faster than `posterior::summarise_draws` for getting rhat and ess?
* Does it make sense for a wrapper to cmdstan to implement a different check from cmdstan by default?

Reply:
Yeah those are all good questions. Initial thoughts (may change upon further reflection): 

> * It's currently the only way to get check the sampler diagnostics and the chain diagnostics (rhat, etc) all at once.

Yeah, I don't think we should deprecate it (if we even do deprecate it even) unless we also have the rhat and ess checks in `fit$diagnose()` too. But that should be easy to do. 

> * Is `cmdstan_diagnose()` nontrivially faster than `posterior::summarise_draws` for getting rhat and ess?

Not sure, but that's definitely something to check before deciding to deprecate it. If it's much faster then maybe we don't deprecate it we just mention in the doc that we recommend the R version unless speed is a concern and the CSV files are also available. 

> * Does it make sense for a wrapper to cmdstan to implement a different check from cmdstan by default?

I think it's ok but not ideal. Ideally we'll change it in CmdStan too if the e-bfmi check is too conservative.